shell/common/node_bindings.cc

```diff
@@ -575,17 +575,27 @@ void NodeBindings::UvRunOnce() {
   // Enter node context while dealing with uv events.
   v8::Context::Scope context_scope(env->context());
 
-  // Perform microtask checkpoint after running JavaScript.
-  gin_helper::MicrotasksScope microtasks_scope(env->isolate());
+  // Deal with uv events.
+  int r = 0;
+  do {
+    // Perform microtask checkpoint after running JavaScript.
+    gin_helper::MicrotasksScope microtasks_scope(env->isolate());
 
-  if (browser_env_ != BrowserEnvironment::kBrowser)
-    TRACE_EVENT_BEGIN0("devtools.timeline", "FunctionCall");
+    if (browser_env_ != BrowserEnvironment::kBrowser)
+      TRACE_EVENT_BEGIN0("devtools.timeline", "FunctionCall");
 
-  // Deal with uv events.
-  int r = uv_run(uv_loop_, UV_RUN_NOWAIT);
+    r = uv_run(uv_loop_, UV_RUN_NOWAIT);
 
-  if (browser_env_ != BrowserEnvironment::kBrowser)
-    TRACE_EVENT_END0("devtools.timeline", "FunctionCall");
+    if (browser_env_ != BrowserEnvironment::kBrowser)
+      TRACE_EVENT_END0("devtools.timeline", "FunctionCall");
+
+    // Make sure that `setImmediate` yields to browser to avoid starvation
+    auto* immediate_idle_handle =
+        reinterpret_cast<uv_handle_t*>(env->immediate_idle_handle());
```

Comment:
Is there a public API that Node.js could reasonable introduce at some point here (without exposing the idle handle itself, which is really more of an implementation detail)?

Reply:
> I think if you want to upstream this, it might be nicer to move the `if (size == 0) { ... }` block in this function to after the `if (popped_value) { ... }` block (and thus avoiding the call to `uv_idle_stop()` if it’s not necessary), instead of restarting the idle handle even if there are not actually any new items in the queue?

Here's the reason why this change is actually needed. In half of the cases that I've observed in our app the `Push` was called from within the JS callback of the threadsafe function. In another half - from the microtasks queue (because there were tons of promises used in JS). The extra iteration is extra cost, but for Electron the cost of not doing the iteration is substantially higher. I'm going to add `else if (!popped_value) {` before `uv_idle_stop()` to simplify the logic, but I think I'd rather have the extra iteration. With this in mind, do you think I should submit it to Node.js?

Reply:
Sounds like a good idea to me. This PR just wants to know if `setImmediate` is scheduled for run in order to interleave `uv_run()` with browser's event loop and avoid starvation.